frame/merkle-mountain-range/src/primitives.rs

```diff
@@ -0,0 +1,413 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use frame_support::{
+	RuntimeDebug,
+	traits::Get,
+	weights::Weight,
+};
+use sp_runtime::traits;
+use sp_std::fmt;
+#[cfg(not(feature = "std"))]
+use sp_std::prelude::Vec;
+
+/// A provider of the MMR's leaf data.
+pub trait LeafDataProvider {
+	/// A type that should end up in the leaf of MMR.
+	type LeafData: FullLeaf;
+
+	/// The method to return leaf data that should be placed
+	/// in the leaf node appended MMR at this block.
+	///
+	/// This is being called by the `on_initialize` method of
+	/// this pallet at the very beginning of each block.
+	/// The second return value should indicate how much [Weight]
+	/// was required to retrieve the data.
+	fn leaf_data() -> (Self::LeafData, Weight);
+}
+
+impl LeafDataProvider for () {
+	type LeafData = ();
+
+	fn leaf_data() -> (Self::LeafData, Weight) {
+		((), 0)
+	}
+}
+
+impl<T: frame_system::Trait> LeafDataProvider for frame_system::Module<T> {
+	type LeafData = <T as frame_system::Trait>::Hash;
+
+	fn leaf_data() -> (Self::LeafData, Weight) {
+		let hash = Self::parent_hash();
+		(hash, T::DbWeight::get().reads(1))
+	}
+}
+
+
+/// A full leaf content stored in the offchain-db.
+pub trait FullLeaf: Clone + PartialEq + fmt::Debug + codec::Decode {
+	/// Encode the leaf either in it's full or compact form.
+	///
+	/// NOTE the encoding returned here MUST be `Decode`able into `FullLeaf`.
+	fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F, compact: bool) -> R;
+}
+
+impl<T: codec::Encode + codec::Decode + Clone + PartialEq + fmt::Debug> FullLeaf for T {
+	fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F, _compact: bool) -> R {
+		codec::Encode::using_encoded(self, f)
+	}
+}
+
+/// An element representing either raw leaf data or a hash.
+///
+/// See [Compact] to see how it may be used in practice to reduce the size
+/// of proofs in case multiple [LeafDataProvider]s are composed together.
+///
+/// [DataOrHash::hash] method is calculating the hash of this element in it's compact form,
+/// so should be used instead of hashing the encoded form (which will always be non-compact).
+#[derive(RuntimeDebug, Clone, PartialEq)]
+pub enum DataOrHash<H: traits::Hash, L> {
+	/// A terminal leaf node, storing arbitrary content.
+	Data(L),
+	/// An inner node - always just a hash.
```

Comment:
I am a bit confused. Does this variant represent only an inner node? I assume not, looking at your example of `CompactLeaf` at the very least.